parquet-hadoop/src/main/java/org/apache/parquet/hadoop/codec/NonBlockedCompressor.java

```diff
@@ -100,7 +102,14 @@ public synchronized void setInput(byte[] buffer, int off, int len) {
         !outputBuffer.hasRemaining(), "Output buffer should be empty. Caller must call compress()");
 
     if (inputBuffer.capacity() - inputBuffer.position() < len) {
-      ByteBuffer tmp = ByteBuffer.allocateDirect(inputBuffer.position() + len);
+      final int newBufferSize;
+      if (inputBuffer.capacity() == 0) {
+        newBufferSize = Math.max(INITIAL_INPUT_BUFFER_SIZE, len);
+      } else {
+        newBufferSize = Math.max(inputBuffer.position() + len, inputBuffer.capacity() * 2);
```

Comment:
Should we set an upper bound to it instead of blindly doubling the capacity? In the new code, we may see much larger peak memory compared to the past.

Reply:
Some analysis:

With doubling, peak memory usage could be up to about 1x the size of the amount of really required memory (if we really needed `x`, and started with `x-1`, we would double to `2x-2`, and then we'd have allocated `x - 2` of unnecessary memory).

If the target size is 64MB (the abnormally large size that I encountered in the wild), starting at 4KB and doubling gets us there in 14 iterations, allocating and deallocating 134MB of total memory.

We could set an upper bound for each allocation at 1MB, so peak memory usage would be at most 1MB more than the amount of really required memory. If we start at 4KB and double up to 1MB, then go in 1MB chunks, we get there in 71 iterations, allocating and deallocating 2GB of total memory.

We could also use `* 1.2` instead of `* 2`, which would make the peak memory usage at most 20% of the amount of really required memory. Starting at 4KB and increasing by 20% each allocation gets us there in 53 iterations, allocating and deallocating 380MB of total memory.

Perhaps 20% growth is a good balance, since it still gets us to target pretty quickly compared to using a 1MB cap, and peak memory usage is at most 20% higher than what is really needed. Please let me know what you think.

Reply:
Thanks for the through analysis! `* 1.2` sounds reasonable to me.  Did you have the time spent on different strides?

Reply:
I haven't measured it, but probably the time spent is a function of the number of iterations and the total amount of memory allocated and deallocated. Compared to what I was seeing without any minimum-increase factor at all, `* 1.2` and `* 2` are both really big improvements.

I just changed the patch to do `* 1.2`.